ssl/ech/ech_internal.c

```diff
@@ -12,4 +12,183 @@
 #include "../ssl_local.h"
 #include "ech_local.h"
 
-/* TODO(ECH): move ECH internal code here when we get to it */
+#ifndef OPENSSL_NO_ECH
+
+/* ECH internal API functions */
+
+static OSSL_ECHSTORE_ENTRY *ossl_echstore_entry_dup(OSSL_ECHSTORE_ENTRY *orig)
+{
+    OSSL_ECHSTORE_ENTRY *ret = NULL;
+    OSSL_ECHEXT *ext = NULL, *newext = NULL;
+    int i;
+
+    if (orig == NULL)
+        return NULL;
+    ret = OPENSSL_zalloc(sizeof(*ret));
+    if (ret == NULL)
+        return NULL;
+    ret->version = orig->version;
+    if (orig->public_name != NULL) {
+        ret->public_name = OPENSSL_strdup(orig->public_name);
+        if (ret->public_name == NULL)
+            goto err;
+    }
+    ret->pub_len = orig->pub_len;
+    if (orig->pub != NULL) {
+        ret->pub = OPENSSL_memdup(orig->pub, orig->pub_len);
+        if (ret->pub == NULL)
+            goto err;
+    }
+    ret->nsuites = orig->nsuites;
+    ret->suites = OPENSSL_memdup(orig->suites, sizeof(OSSL_HPKE_SUITE) * ret->nsuites);
+    if (ret->suites == NULL)
+        goto err;
+    ret->max_name_length = orig->max_name_length;
+    ret->config_id = orig->config_id;
+    if (orig->exts != NULL) {
```

Comment:
Can you use sk_OSSL_ECHEXT_deep_copy()  here?

Reply:
no idea tbh, hadn't seen `_deep_copy()` before, doesn't look like it'd result is less code really and would maybe be less readable code for something we don't expect to see really used, so am a bit reluctant...

Reply:
What drew my attention to it is the size of this function..
This would make this actual function smaller by shifting the work to another function.

The code would look something like this...
````
static void echext_free(const OSSL_ECHEXT *ext)
{
    OPENSSL_free(newext->val);
    OPENSSL_free(newext);
}

static OSSL_ECHEXT *echext_copy(const OSSL_ECHEXT *src)
{
   OSSL_ECHEXT *ext = OPENSSL_zalloc(sizeof(*src));
 
   if (ext != NULL)
      return NULL;
   *ext = *src;
   ext->val = NULL;
   if (ext->len != 0) {
         ext->val = OPENSSL_memdup(src->val, src->len);
          if (ext->val == NULL)  {
              echext_free(ext);
              return NULL;
           }
      }
      return ext;     
 }

ret->exts = sk_OSSL_ECHEXT_deep_copy(orig->exts, echext_copy, echext_free);
````

Reply:
well... willing to do it, sure, though it seems a tiny bit (but only tiny) less transparent, but if that's what you guys prefer to maintain, that wins the argument, will push in a sec (possibly twice, when I realise I did the 1st one wrong:-)

Reply:
Looks like that worked ok (modulo the external oqs test that's been failing for a while, presumably fixable via a rebase). Thanks for the sample code though, made it much easier.

Reply:
Should be ok to use since this deep_cop[y is used in other places such as X509, cmp, and encoders.